### PR TITLE
Refactors High Luminosity Eyes, fixes a ton of bugs related to it as well as qol improvements

### LINKED ISSUE #14685: High luminosity eyes beam not following owner
## Round ID:
5842

## Testmerges:
[#12529: '[MODULAR] Status Indicators with Preference' by nevimer at commit 360dfe72a7](https://github.com/Skyrat-SS13/Skyrat-tg/pull/12529)
[#10823: 'Database revision revert' by Gandalf2k15 at commit 30f7853899](https://github.com/Skyrat-SS13/Skyrat-tg/pull/10823)
[#13848: 'The Refactaur' by GoldenAlpharex at commit f26740d174](https://github.com/Skyrat-SS13/Skyrat-tg/pull/13848)
[#14593: '[DNM] [TM ONLY] Revert "[MIRROR] The Humanening?! But cheeked up (enlarged posterior) on a Sunday night?!?! As per our sprite lead's review????? [MDB IGNORE]"' by jjpark-kb at commit fdd6a814fe](https://github.com/Skyrat-SS13/Skyrat-tg/pull/14593)
[#14647: 'Update orbit.dm' by Gandalf2k15 at commit 7ea07af42b](https://github.com/Skyrat-SS13/Skyrat-tg/pull/14647)
[#14625: 'Railroads' by jjpark-kb at commit 234abb4ffc](https://github.com/Skyrat-SS13/Skyrat-tg/pull/14625)
[#14023: 'Fleshmind [PRE-RELEASE FINAL]' by Gandalf2k15 at commit 6d1a9d2338](https://github.com/Skyrat-SS13/Skyrat-tg/pull/14023)

## Reproduction:
Equip high luminosity eyes, walk in a straight line, have the light not follow.

### LINKED ISSUE #61041: eyelighting only displays non-white color on first tile
## Round ID:

was seen downstream but I believe the bug is upstream

## Reproduction:
Get the light-up eyes
change the color

image

### PR DESCRIPTION
**Pull request by vinylspiders**
## About The Pull Request

The high luminosity eyes item was extremely out of date, broken, and full of copy paste code from atom lighting. Which is a shame because they were cool.

On top of all that it was using a special light effect that was not very performant. I got rid of all that, hooked it into atom lighting as a new light type, and gave it a new TGUI as well because the old ui prompts were not great either. 

You can now pick a color at random if you want. You can also set the color and range before surgically implanting them. No more being forced to go through the color picker when you just want to change the range.

Functionally they should pretty much should be the same as before with some bonus features (see below).

dreamseeker_nDeLNyOOG2

Closes https://github.com/tgstation/tgstation/issues/61041
Closes https://github.com/Skyrat-SS13/Skyrat-tg/issues/14685

This is 100% completed. I just finished fixing the slight translation bug when going from 0->1 range (see above gif) and that was the last thing on my bucket list. I happy enough with this at this point in time.

---

EDIT: 

I have decided to add in one last new feature, and that is... independent settings for eye color.

<details> <summary>You can now set eye color independently if you wish</summary>

dreamseeker_j32B2S4yXQ

</details>

The eye color does not modify the light color in any way when set in this manner, but it can be used for cosmetic purposes. 

Kind of makes the item more like cybereyes from cyberpunk, which I think are pretty neat!

</details>

### What I've done, in more detail:

- refactored high luminosity eyes so they use the atom lighting system instead of the way they were doing it before
- the new light type, `MOVABLE_LIGHT_BEAM` behaves similarly to directional lights, with some slight differences. it reuses the same lighting overlay sprites but scales them vertically to produce a more focused effect. The result can be seen above. This is in contrast to the old way, which spawned `range` number of individual 32x32 overlays and moved them around. This way should perform better as well as be more maintainable.
- added a new TGUI interface for high luminosity eyes with buttons for range, a text field for a color hex, a color picker and randomizer
- made the eye overlay emissive when the light is turned on
- range goes from 0 to 5. at range 0, the light overlay is turned off and you are left with just the emissive eyes.
- added a cosmetic functionality to this item that lets you change the color of your eyes independently of the lighting (and each other)
- fixed a bug with directional flashlights  sometimes not updating their lighting overlay if you pick them up without changing your direction
---

### Other Misc Fixes

Being able to dynamically set range back and forth exposed some logic issues that had existed with directional light overlays and I have fixed those. That is why there are some edits in that file that may not appear readily obvious why they are there.

Apart from that, two other bugs that come to mind:
1) eye code was supposed to keep track of the eye color you had before you got new eyes, but it was overwriting that every time you ran refresh(). 
2) lighting was supposed to be turning off the light when range is set to 0, but it was not doing that properly.

And of course besides that, there may have been a few instances of finding typos/tidying up code

## Why It's Good For The Game

The code for this was like 6 years old and in desperate need of updating. Now it works, and has a nicer UI.

## Changelog

:cl:
fix: high luminosity eyes light overlays now follow the user correctly
qol: high luminosity eyes now have a tgui menu so you no longer have to go through the color picker every time you want to change the range. they also have a new setting that lets you change the color of your eyes independently of the light color. You can now have cybernetic heterochromia if you want
fix: directional flashlights when picked up will now always update their cast light direction correctly no matter what dir you are facing
refactor: refactors high luminosity eye code to better make use of the atom lighting system, adding a new light type 'MOVABLE_LIGHT_BEAM'
/:cl:

### CODE REVIEW

**Comment by san7890** [comment]
@LemonInTheDark lighting stuff, thoughts?

**Comment by larentoun** [comment]
After fastly fixing the issue with "moving with owner", I was very tempted to try to make these eyes newer (like here). I'm so glad that I didn't start it

File: tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+import { Window } from '../layouts';
+
+type EyeColorData = {
+  mode: BooleanLike;
+  hasOwner: BooleanLike;
+  left: string;
+  right: string;
+};
+
+type HighLuminosityEyesMenuData = {
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx:13]: i appreciate this effort but it's a bit lengthy for a local type name

"Data" is fine

File: tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+  eyeColor: EyeColorData;
+  lightColor: string;
+  range: number;
+};
+
+const LightColorDisplay = (props, context) => {
+  const { act, data } = useBackend<HighLuminosityEyesMenuData>(context);
+  const { lightColor } = data;
+  const light = 0;
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx:22]: this should be made a global constant or an enum

File: tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+      </LabeledList>
+    </Section>
+  );
+};
+
+const EyeColorDisplay = (props, context) => {
+  const { act, data } = useBackend<HighLuminosityEyesMenuData>(context);
+  const { eyeColor } = data;
+  const left = 1;
+  const right = 2;
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx:93]: these should also be an enum, easy case here,
`enum Eye {
 Left,
 Right
}`

File: tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+export const HighLuminosityEyesMenu = (props, context) => {
+  const { act, data } = useBackend<HighLuminosityEyesMenuData>(context);
+  const { eyeColor } = data;
+  return (
+    <Window
+      title="High Luminosity Eyes"
+      width={262}
+      height={eyeColor.hasOwner ? (eyeColor.mode ? 259 : 324) : 188}>
+      <Window.Content scrollable>
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx:164]: i'm not super thrilled on this design, i think it's a bit redundant to use multiple sections and labeled lists which can be squashed into just one.

I'm a hater of scrollable window content, I think one `section fill scrollable` should replace it

**Review by jlsnow301 — CHANGES_REQUESTED** [review]

File: tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx (hunk 3 above)
Comment (Mothblocks) [reply tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx]: It's identical

File: tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx (hunk 4 above)
Comment (jlsnow301) [reply tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx]: Aesthetic and vendetta against NT :). Sticky headers just looks entirely better to me, especially for navigational components. I hate having to scroll back to top to navigate elsewhere. Basically, this effect:
image

File: tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+  LeftEye,
+  RightEye,
+}
+
+const LightColorDisplay = (props, context) => {
+  const { act, data } = useBackend<Data>(context);
+  const { lightColor } = data;
+  return (
+    <Section title="Color">
+      <LabeledList>
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx:30]: these can be crushed as well, remove section & labeled list

File: tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+    </Section>
+  );
+};
+
+const EyeColorDisplay = (props, context) => {
+  const { act, data } = useBackend<Data>(context);
+  const { eyeColor } = data;
+  return (
+    <Section title="Eye Color">
+      <LabeledList>
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx:106]: and here

File: tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+export const HighLuminosityEyesMenu = (props, context) => {
+  const { act, data } = useBackend<Data>(context);
+  const { eyeColor } = data;
+  return (
+    <Window
+      title="High Luminosity Eyes"
+      width={262}
+      height={eyeColor.hasOwner ? (eyeColor.mode ? 259 : 324) : 188}>
+      <Section fill scrollable>
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx:186]: add labeled list here pls

File: tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+          <Button.Checkbox
+            checked={eyeColor.mode}
+            onClick={() => act('toggle_eye_color')}
+            tooltip="Toggle the eye color mode."
+          />
+        </LabeledList.Item>
+      </LabeledList>
+      {!eyeColor.mode && (
+        <LabeledList>
+          &nbsp;
```
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx:117]: move to main component please

File: tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx (hunk 4 above)
Comment (jlsnow301) [reply tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx]: looking at this I think I didn't communicate that window.content was still okay to leave in. i'll message you about this on discord, the scrollable prop goes on section, i meant

File: tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx (hunk 3 above)
Comment (jlsnow301) [reply tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx]: even using a const-as-const would be okay here but the codebase has kind of settled as using enums for interfacing js-to-dm.
however, even left as a const, it's better to move things to top level scope, as the component is recreating it on each render. in this case it's barely a micro op but more of a convention. somewhat related:

https://stackoverflow.com/questions/80802/does-use-of-anonymous-functions-affect-performance/81329#81329

the general rule is moving everything to the top level that doesn't explicitly depend on or reference a state etc.

File: tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx (hunk 4 above)
Comment (jlsnow301) [reply tgui/packages/tgui/interfaces/HighLuminosityEyesMenu.tsx]: Here's a version I put together rq that I think embodies minimizing redundant section/labels
dreamseeker_tnrB1syglL

**Pull request merged**